expr/expr.go

```diff
@@ -33,6 +33,115 @@ func Unmarshal(fam byte, data []byte, e Any) error {
 	return e.unmarshal(fam, data)
 }
 
+// ExprsFromBytes parses nested raw expressions bytes
+// to construct nftables expressions
+func ExprsFromBytes(fam byte, ad *netlink.AttributeDecoder, b []byte) ([]Any, error) {
+	var exprs []Any
+	ad.Do(func(b []byte) error {
+		ad, err := netlink.NewAttributeDecoder(b)
+		if err != nil {
+			return err
+		}
+		ad.ByteOrder = binary.BigEndian
+		var name string
+		for ad.Next() {
+			switch ad.Type() {
+			case unix.NFTA_EXPR_NAME:
+				name = ad.String()
+				if name == "notrack" {
+					e := &Notrack{}
+					exprs = append(exprs, e)
+				}
+			case unix.NFTA_EXPR_DATA:
+				var e Any
+				switch name {
+				case "ct":
+					e = &Ct{}
+				case "range":
+					e = &Range{}
+				case "meta":
+					e = &Meta{}
+				case "cmp":
+					e = &Cmp{}
+				case "counter":
+					e = &Counter{}
+				case "payload":
+					e = &Payload{}
+				case "lookup":
+					e = &Lookup{}
+				case "immediate":
+					e = &Immediate{}
+				case "bitwise":
+					e = &Bitwise{}
+				case "redir":
+					e = &Redir{}
+				case "nat":
+					e = &NAT{}
+				case "limit":
+					e = &Limit{}
+				case "quota":
+					e = &Quota{}
+				case "dynset":
+					e = &Dynset{}
+				case "log":
+					e = &Log{}
+				case "exthdr":
+					e = &Exthdr{}
+				case "match":
+					e = &Match{}
+				case "target":
+					e = &Target{}
+				case "connlimit":
+					e = &Connlimit{}
+				}
+				if e == nil {
+					// TODO: introduce an opaque expression type so that users know
+					// something is here.
+					continue // unsupported expression type
+				}
+
+				ad.Do(func(b []byte) error {
+					if err := Unmarshal(fam, b, e); err != nil {
+						return err
+					}
+					// Verdict expressions are a special-case of immediate expressions, so
+					// if the expression is an immediate writing nothing into the verdict
+					// register (invalid), re-parse it as a verdict expression.
+					if imm, isImmediate := e.(*Immediate); isImmediate && imm.Register == unix.NFT_REG_VERDICT && len(imm.Data) == 0 {
+						e = &Verdict{}
+						if err := Unmarshal(fam, b, e); err != nil {
+							return err
+						}
+					}
+					exprs = append(exprs, e)
+					return nil
+				})
+			}
+		}
+		return ad.Err()
+	})
+	return exprs, ad.Err()
+}
+
+// ExprsFromMsg parses netlink message in order to
+// extract and construct expressions
+func ExprsFromMsg(fam byte, b []byte) ([]Any, error) {
```

Comment:
It’s not clear to me why these two functions need to be exported? Can you clarify please?

Reply:
The `ExprsFromMsg` was previously [`exprsFromMsg` in `rule.go`](https://github.com/google/nftables/blob/3e042f75d78f94b3182304566f908f5bf7d7a7ad/rule.go#L218), but since it does not make sense to implement the function twice (it is needed to parse dynset expressions the same way as rule expressions) it was moved to the `expr` package and exported to be able to use it in both `rule.go` and `dynset.go`.

Reply:
I’m concerned about exporting these and thereby making them part of the public API. Can you move the functionality into a package within the `internal/` directory instead please?

Reply:
@stapelberg Understandable. I would be very happy to do it but unfortunately I am not sure how to move it without redesigning current source code placement.

The problem with moving the parsing functionality outside of the `expr` package is in the cyclic dependency that it creates between `expr` and `internal`.

Since the parsing logic extensively uses functions and structs from the `expr` package (`Unmarshal`, `Marshal`, `&expr.Notrack`, ...) moving it out of `expr` package would create the need to reference this (now moved) implementation from the same `expr` package while the implementation references `expr` package structs and functions. 

Let me try to depict this "graphically":
```
expr/dynset.go
  ^       |
  |       |
._|_________________________________________.
|uses expr structs and functions for parsing|
._|_________________________________________.
  |
  | CYCLE |
  |    .__|_____________________________________________________.
  |    |needs parsing_implementation to parse dynset expressions|
  |    .__|_____________________________________________________.
  |       |
  |       |
  |       v
internal/nfexpr/parsing_implementation.go
    ^
    |
  ._|____________________________________________________.
  |needs parsing_implementation to parse rule expressions|
  ._|____________________________________________________.
    |
    |
rule.go
```

I could theoretically overcome this by putting a part of the parsing implementation (the one which is not common for both `rule.go` and `dynset.go`?) in the newly created `internal` package and then force the `dynset.go` implementation to reference the parsing implementation in the `expr` package and `rule.go` to reference the parsing implementation in the newly created `internal` package where afterwards both reference the same (common) parsing implementation in the `expr` package, but this would still require me to export this parsing implementation in `expr` in order for it to be reachable from the newly created `internal` package.

Something like this:
```
expr/dynset.go (part1 parsing in Unmarshal)
         |
      .__|_____________________.
      |parse dynset expressions|
      .__|_____________________.
         |
         |
         v
expr/part2_of_parsing_implementation.go
       ^
       |
.______|________________________________________.
|parse rule expressions (cannot call if private)|
.______|________________________________________.
       |
       |
internal/nfexpr/part1_of_parsing_implementation.go
    ^
    |
  ._|____________________.
  |parse rule expressions|
  ._|____________________.
    |
    |
rule.go
```

Thinking twice, this approach does not make sense since the first part of rule parsing can just be written as a private function which would again need to call into `expr`, forcing the implementation in `expr` to be exported. The only advantage in this approach might be in exposing only a smaller part of the parsing implementation that you would not want to export.

Another approach I was thinking about was creating an `internal` package inside `expr` package but this is not possible since then the parsing implementation would be completely unavailable to `rule.go`.

Maybe I am just overlooking something. Let me think about other ways to approach this, but please feel free to ping here if you have any ideas.

Reply:
Yeah, it’s a tricky situation. I think the following strategy could work, though:

1. Introduce a `internal/parseexprfunc` package which contains `var ParseFunc func(fam byte, ad *netlink.AttributeDecoder, b []byte) ([]interface{}, error)` (we need to use the empty interface instead of `expr.Any` to avoid a circular dependency).
2. In package `expr`, add:
    ```
   func init() { 
    parseexprfunc.ParseFunc = func(fam byte, ad *netlink.AttributeDecoder, b []byte) ([]interface{}, error) {
      exprs, err := exprsFromBytes(fam, ad, b)
      if err != nil {
        return nil, err
      }
      result := make([]interface{}, len(exprs))
      for idx, expr := range exprs {
        result[idx] = expr
      }
      return result, nil
    }
    ```
3. From `rule.go`, you can now call `parseexprfunc.ParseFunc()` and convert the opposite direction (from `interface{}` to `expr.Any`), and within package `expr`, you can call the non-exported `exprsFromBytes` for the dynset.

Reply:
@stapelberg Thanks for the suggestion! I am not particularly happy with performing conversions back and forth but since I have no better approach to suggest, I have used your suggestion and implemented the `parseexprfunc` package.